### PR TITLE
Test case to check updated_date

### PR DESCRIPTION
**Pull request by shashank-elastic**
## Issues
https://github.com/elastic/ia-trade-team/issues/118

## Summary

- The purpose of the PR is to enforce the rule authors are updating the rule modified date in field `updated_date` before merging to main via pull_request.

## Additional Information

- The test is not intended  to cover direct commits to release branches, because as part of our back-porting workflows we [unstage-incompatible-rules](https://github.com/elastic/detection-rules/blob/main/.github/workflows/backport.yml#L108) before committing to our release branches. Which means when we do the diff there is a discrepancy between what diff main report and what the release branch intends to have.  Direct commits to release branches should run unit tests locally as a best practise before committing. 
- We added an explicit skip check to check for GitHub events "push" and skip the test case on those events. This will allow local testing without any additional environment variables for dev environments 

## Code Changes 

- Add a Test to ensure modified rules have updated_date field updated.
- The test case will be part of TestRuleMetadata class
- All rules under `rules/` and `rules_building_block` are considered.
- Git diff will fetch us rules modified, and then git diff of each modified file is obtained to fetch the changes
- To handle usecase  when a new rule is modified , we use `'--diff-filter=M'` where only modified files are considered for rule path `rules/` and `rules_building_block/`
- If a rule is modified, then the git diff is examined to check for updated_date changes. 
- If the Updated_date is never modified test case fails.
- If the diff changes does not contain "updated_date", then fail test case
- To handle failures when rule changes are backported to protected version branches explained [here](https://github.com/elastic/ia-trade-team/issues/118#issuecomment-2183036472)  we have added a test case skip criteria 

## Edge Case 
- Now the dev branch is open for x days and rule is tuned and updated_date is updated once, this test case will succeed if the rule is modified and updated_date is left unchanged after the first updation. All it checks is to contain one modification of updated_date. 
- Also if the Branch is merged on a diff date the updated_date can differ from the commit date to main and this will always pass on main. All it checks is to contain one modification of updated_date. 

## Testing

- Updated_date check testing is the same as updated in PR - https://github.com/elastic/detection-rules/pull/3764
- Skip criteria Testing.
Local Testing with no git hub event and base ref --> Test Case Skipped.
```
tests/test_all_rules.py::TestRuleFiles::test_rule_file_name_tactic PASSED                                                                                                                                                        [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules PASSED                                                                                                                                                          [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                                 [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                             [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                                                                                                                                                           [ 60%]
tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date SKIPPED (Skipping this test when not running on pull requests to main branch)             
```
Local Testing with following git hub events  Event- pull_request and Branch = 8.14 --> Test Case Skipped.
```
❯ export GITHUB_EVENT_NAME=pull_request  
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ export GITHUB_BASE_REF=8.14
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_EVENT_NAME
pull_request
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_BASE_REF
8.14
```
```
tests/test_all_rules.py::TestRuleFiles::test_bbr_in_correct_dir PASSED                                                                                                                                                           [ 55%]
tests/test_all_rules.py::TestRuleFiles::test_non_bbr_in_correct_dir PASSED                                                                                                                                                       [ 56%]
tests/test_all_rules.py::TestRuleFiles::test_rule_file_name_tactic PASSED                                                                                                                                                        [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules PASSED                                                                                                                                                          [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                                 [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                             [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                                                                                                                                                           [ 60%]
tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date SKIPPED (Skipping this test when not running on pull requests to main branch)  
```

Local Testing with following git hub events  Event- push and Branch = main--> Test Case Skipped.
```
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ export GITHUB_EVENT_NAME=push        
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ export GITHUB_BASE_REF=main   
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_EVENT_NAME       
push
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_BASE_REF         
main
```

```
tests/test_all_rules.py::TestRuleFiles::test_bbr_in_correct_dir PASSED                                                                                                                                                           [ 55%]
tests/test_all_rules.py::TestRuleFiles::test_non_bbr_in_correct_dir PASSED                                                                                                                                                       [ 56%]
tests/test_all_rules.py::TestRuleFiles::test_rule_file_name_tactic PASSED                                                                                                                                                        [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules PASSED                                                                                                                                                          [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                                 [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                             [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                                                                                                                                                           [ 60%]
tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date SKIPPED (Skipping this test when not running on pull requests to main branch)       
```

Local Testing with following git hub events  Event- pull_request and Branch = main--> Test Case Executed
```
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co took 1m23s 
❯ export GITHUB_EVENT_NAME=pull_request
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_EVENT_NAME              
pull_request
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_BASE_REF                
main
```

```
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                                 [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                             [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                                                                                                                                                           [ 60%]
tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date PASSED                                                                                                                                              [ 61%]
tests/test_all_rules.py::TestRuleMetadata::test_updated_date_newer_than_creation PASSED          
```

On Github PR Run. The testcase was executed - [sample_run](https://github.com/elastic/detection-rules/actions/runs/9663231624/job/26654898511?pr=3818)

image

#### Testing With Skip criteria for push

Locally this can be run without any additional settings

```console
tests/test_all_rules.py::TestRuleTimelines::test_timeline_has_title PASSED                                                                                                                                                     [ 55%]
tests/test_all_rules.py::TestRuleFiles::test_bbr_in_correct_dir PASSED                                                                                                                                                         [ 55%]
tests/test_all_rules.py::TestRuleFiles::test_non_bbr_in_correct_dir PASSED                                                                                                                                                     [ 56%]
tests/test_all_rules.py::TestRuleFiles::test_rule_file_name_tactic PASSED                                                                                                                                                      [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules PASSED                                                                                                                                                        [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                               [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                         [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                        
```

On PR the test is checked --> Refer Latest Unit Test Result.

When event is push the test case is skipped
```console
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co took 1m25s 
❯ export GITHUB_EVENT_NAME=push          
(.venv) 
detection-rules on  tarde_118 [$?] is 📦 v0.1.0 via 🐍 v3.12.3 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ echo $GITHUB_EVENT_NAME       
push

tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules PASSED                                                                                                                                                        [ 57%]
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules_modified PASSED                                                                                                                                               [ 58%]
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset PASSED                                                                                                                                                           [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag PASSED                                                                                                                                                         [ 59%]
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries PASSED                                                                                                                                                         [ 60%]
tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date SKIPPED (Skipping this test when not running on pull requests.)       
```

### CODE REVIEW

File: .github/workflows/pythonpackage.yml
Comment (Mikaayenson) [inline .github/workflows/pythonpackage.yml:46]: I can't tell how this will only run the test case on main vs only running the entire step (all unit tests) only on main.  We essentially need all unit tests to run on our release branches, so if it does limit unit tests to run on only the main branch, then we should rethink this.

Comment (shashank-elastic) [reply .github/workflows/pythonpackage.yml:46]: At this point we are not at all limiting test case running we are just pushing 2 env variables for pytest usages. 
The actual skip is on this line. https://github.com/elastic/detection-rules/pull/3818/files#diff-143495f0794a3e8b322e5cbf9a4735acab2213fd6d0f678149b78d322976507bR643

If you look at the local testing where we have simulated running on release branches you see all test cases executed and this being skipped

File: tests/test_all_rules.py
Comment (Mikaayenson) [inline tests/test_all_rules.py:644]:
```suggestion
    @unittest.skipIf(os.getenv('GITHUB_BASE_REF', 'main') != 'main',
```

File: tests/test_all_rules.py
Comment (Mikaayenson) [inline tests/test_all_rules.py:645]:
```suggestion
```

We would still want to be able to test locally before opening a PR.

File: tests/test_all_rules.py
Comment (Mikaayenson) [inline tests/test_all_rules.py:658]: ? Is it possible to diff based on the dynamic branch so that it runs appropriate on main and the backport branches?

**Comment by shashank-elastic** [comment]
Reconciling this PR design

- The skip test on non pull requests was added to ensure this test would be skipped when we push the backport commits. 
- Edge Case [like](https://github.com/elastic/ia-trade-team/issues/118#issuecomment-2183036472) this can be tackled if this pull requests runs once only on main branch was the initial consideration for this PR.
- With this comes the problem of having certain variables set to be able to run it locally. [Refer the testing criteria]
- What are the possibilities that we can have without skipping
    -- We could dynamically change the base branch ref on PR / Push event to  use those. 
    -- My Point of view on this is that the test on push events is invariably wont fetch any events. 
    -- There is a diff of the rule on the PR stage, on Push events ur already looking at the modified file and comparing it with its base branch wont fetch any diff. 
    -- On the other hand, back porting branches wont have any PR(s) and are always on direct commit mode. 
    -- So even if we have this test case to have diff on dynamic branch there will be no result as u will be diffing on the latest branch contents. 
 - So we could use a simple negate and skip if the event type is "push" and rest of the cases the test case will run and we will be able to execute this seamlessly locally as well. 
 
 opening this up to @Mikaayenson / @eric-forte-elastic / @traut to see if this is less disruptive for dev workflows.

File: tests/test_all_rules.py
Comment (shashank-elastic) [reply tests/test_all_rules.py:658]: See new design discussion

File: tests/test_all_rules.py
Comment (eric-forte-elastic) [inline tests/test_all_rules.py:650]:
```suggestion
        rules_path = get_path("rules")
        rules_bbr_path = get_path("rules_building_block")
```
This looks good, the rest if this comment is a NOP until this is merged.
Note for our future selves: we will need to merge this in with DAC to use the rules config. Using this comment as to anchor reference.

**Review by eric-forte-elastic — APPROVED** [review]
:green_circle: Manual review, great testing. Looks good to me! :+1:

File: tests/test_all_rules.py
Comment (eric-forte-elastic) [inline tests/test_all_rules.py:655]:
```suggestion
        # Use git diff to check if the file(s) has been modified in rules/ rules_build_block/ directories.
        # For now this checks even rules/_deprecated any modification there will fail
        # the test case "test_deprecated_rules_modified", which means an ignore directory
        # is not required as there is a specific test for deprecated rules.
```

Nit grammar updates.

**Review by traut — APPROVED** [review]
the approach looks good, approved! I do not fully understand all the git flow black magic there but the purpose is clear

File: tests/test_all_rules.py
Comment (Mikaayenson) [inline tests/test_all_rules.py:645]: Only concern I had with this is that we won't be able to naturally run unit tests without setting this environment variable. Probably not ideal for people (internally or the community) that want to test everything locally before opening the PR.

**Review by Mikaayenson — COMMENTED** [review]
You can _probably_ test this on a release branch to try to replicate the error that appeared last time you tried to merge. This will at least tell you if it'll fail again or not.

File: tests/test_all_rules.py
Comment (shashank-elastic) [reply tests/test_all_rules.py:645]: Refer to the test case "Locally this can be run without any additional settings". If we dont have the env variable set which necessarily means its not push and we can run on local settings. The concern is addressed @Mikaayenson

Comment (shashank-elastic) [reply tests/test_all_rules.py:645]: Updated summary

**Review by Mikaayenson — APPROVED** [review]
LGTM. Makes sense with the summary update and change to the skip condition.

**Pull request merged**
